r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java

```diff
@@ -697,6 +751,23 @@ private void timeoutObjects()
     return toReap;
   }
 
+  private <U> Collection<TimeTrackingCallback<U>> reapWaiters(Queue<TimeTrackingCallback<U>> queue, long timeout)
+  {
+    List<TimeTrackingCallback<U>> toReap = new ArrayList<>();
+    long now = _clock.currentTimeMillis();
+    long target = now - timeout;
+
+    synchronized (_lock)
+    {
+      for (TimeTrackingCallback<U> p; (p = queue.peek()) != null && p.getTime() < target;)
+      {
+        toReap.add(queue.poll());
+        _statsTracker.incrementWaiterTimedOut();
```

Comment:
We can move the increment on line 731.
`        obj.onError(new WaiterTimeoutException("Waiter timeout after " + _waitTimeout + "ms"));`

Reply:
The reason that is not done is to avoid acquiring the lock again just for updating the status (similar to timeoutObjects method). Since any slight increase in wait in this class is in the critical path of the request - we should avoid that. Let me know.